File: app/src/main/java/com/example/walmartpathfinder/models/ConcreteSection.kt
package com.example.walmartpathfinder.models

class ConcreteSection(name: String, parent: Section? = null) : Section(name, parent) {
    private var subsections: MutableList<Section> = mutableListOf<Section>()


    /*Allows user to add subsections to subsections list*/
    fun addSubsection(subsection: Section) {
        subsections.add(subsection)
        subsection.setParent(this)
    }

    fun getSubsections(): MutableList<Section> {
        return this.subsections
    }

    /* Check to see the item being checked is a descendant of the specific section */
    override fun sortItemList(unorderedList: MutableList<Item>): Pair<MutableList<Item>, MutableList<Item>> {
        var unordered = mutableListOf<Item>()
        var ordered = mutableListOf<Item>()

        for (item in unorderedList) {
            if (item.isDescendantOf(this)) {
                ordered.add(item)
            } else {
                unordered.add(item)
            }

        }

        /* Add all items back to ordered and unordered list and return both lists */
        var subsectionUnordered = ordered
        ordered = mutableListOf()
        for (subsection in this.subsections) {
            var listPair = subsection.sortItemList(subsectionUnordered)
            ordered.addAll(listPair.first)
            subsectionUnordered = listPair.second
        }

        return Pair(ordered, unordered)
    }

}

File: src/models/ConcreteSection.ts
import { Section } from "./Section"
import { Item } from './Item'

class ConcreteSection extends Section{
    subsections: Array<Section> = [];

    constructor(name: string, parent: Section | undefined){
        super();
        this.name = name;
        this.parent = parent ;
    }

    addSubsection(subsection: Section){
        this.subsections.push(subsection);
        subsection.setParent(this);
    }

    getSubsections(): Array<Section>{
        return this.subsections;
    }

    sortItemList(unorderedList: Array<Item>): [Array<Item>, Array<Item>] {
        // console.log(`Concrete Section input: ${this.name}`)
        // console.log(unorderedList)
        let unordered: Array<Item> = [];
        let ordered: Array<Item> = [];

        for (let i = 0; i < unorderedList.length; i++){
            const item = unorderedList[i]
            if (item.isDescendantOf(this)){
                ordered.push(item)
            }else{
                unordered.push(item);
            }
        }

        let subsectionUnordered = ordered;
        ordered = [];
        for (let i = 0; i < this.subsections.length; i++){
            const subsection = this.subsections[i]
            let listPair = subsection.sortItemList(subsectionUnordered)
            ordered = ordered.concat(listPair[0])
            subsectionUnordered = listPair[1]
            if (subsectionUnordered.length == 0) {
                break;
            }
        }
        // console.log(`Concrete Section return: ${this.name}`)
        // console.log([ordered, unordered])
        return [ordered, unordered]
    }
}

export {ConcreteSection}